mini-projects/01-device-driver/laptop-interface/src/controller.rs

```diff
@@ -0,0 +1,516 @@
+                            let received_data =
+                                String::from_utf8_lossy(&buffer[..bytes_read]).to_string();
+                            log::trace!(
+                                "Controller {}: Received raw data: {:?}",
+                                controller_id,
+                                &buffer[..bytes_read]
+                            );
+                            if let Err(e) = thread_sender.send((controller_id, received_data)) {
+                                log::error!(
+                                    "Controller {}: Failed to send received data: {}",
+                                    controller_id,
+                                    e
+                                );
+                                log::info!(
+                                    "Controller {}: Breaking loop due to send failure.",
+                                    controller_id
+                                );
+                                // Break the loop if sending fails
+                                break;
+                            }
+                            log::trace!(
+                                "Controller {}: Received data processed and sent.",
+                                controller_id
+                            );
+                        } else {
+                            log::trace!(
+                                "Controller {}: No data read from serial port.",
+                                controller_id
+                            );
+                        }
+                    }
+                    Err(e) => match e {
+                        SerialError::Timeout => {
+                            log::debug!(
+                                    "Controller {}: Read operation timed out (This is normal when there is no data to be read)",
+                                    controller_id
+                                );
+                        }
+                        _ => {
+                            log::error!(
+                                "Controller {}: Error reading from serial port: {:?}",
+                                controller_id,
+                                e
+                            );
+                            break;
+                        }
+                    },
+                }
+                log::trace!(
+                    "Controller {}: Serial port read operation completed.",
+                    controller_id
+                );
+
+                log::debug!("Controller {}: End of loop iteration.", controller_id);
+                thread::sleep(Duration::from_millis(100));
+            }
+
+            log::warn!("Controller {}: Exiting thread loop.", controller_id);
+
+            Ok(())
+        }))
+    }
+
+    fn send_message_to_controller(&self, id: u32, message: String) -> Result<(), ControllerError> {
+        if let Some(controller) = self.controllers.iter().find(|c| c.id == id) {
```

Comment:
might've been worth making `controllers` a hash map so that you don't need to constantly do this lookup